ldelf/ta_elf.c

```diff
@@ -855,6 +861,74 @@ static void populate_segments(struct ta_elf *elf)
 	}
 }
 
+static void ta_elf_add_bti(struct ta_elf *elf)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct segment *seg = NULL;
+	uint32_t flags = LDELF_MAP_FLAG_EXECUTABLE | LDELF_MAP_FLAG_BTI;
+
+	TAILQ_FOREACH(seg, &elf->segs, link) {
+		vaddr_t va = elf->load_addr + seg->vaddr;
+
+		if (seg->flags & PF_X) {
+			res = sys_set_prot(va, seg->memsz, flags);
+			if (res)
+				err(res, "sys_set_prot");
+		}
+	}
+}
+
+static void parse_property_segment(struct ta_elf *elf)
+{
+	char *desc = NULL;
+	size_t align = elf->prop_align;
+	size_t desc_offset = 0;
+	size_t prop_offset = 0;
+	vaddr_t va = elf->load_addr + elf->prop_start;
+	Elf_Note *note = (void *)va;
+	char *name = (char *)(note + 1);
+
+	if (elf->prop_memsz < sizeof(*note) + sizeof(ELF_NOTE_GNU))
+		return;
+
+
+	if (note->n_type != NT_GNU_PROPERTY_TYPE_0 ||
+	    note->n_namesz != sizeof(ELF_NOTE_GNU) ||
+	    memcmp(name, ELF_NOTE_GNU, sizeof(ELF_NOTE_GNU)))
+		return;
+
+	desc_offset = ROUNDUP(sizeof(*note) + sizeof(ELF_NOTE_GNU), align);
+
+	if (desc_offset > elf->prop_memsz ||
+	    ROUNDUP(desc_offset + note->n_descsz, align) > elf->prop_memsz)
+		return;
+
+	desc = (char *)(va + desc_offset);
+
+	do {
+		Elf_Prop *prop = (void *)(desc + prop_offset);
+		size_t data_offset = prop_offset + sizeof(*prop);
+
+		if (note->n_descsz < data_offset)
+			return;
+
+		if (prop->pr_type == GNU_PROPERTY_AARCH64_FEATURE_1_AND) {
+			uint32_t *pr_data = (void *)(desc + data_offset);
```

Comment:
`data_offset` is read from the ELF file and seems used as an index used to read memory. Possibly subject to speculative unsafe accesses. Use `confine_array_index()` to ensure `data_offset` is inside TA range?

Reply:
Added